src/librustdoc/html/highlight.rs

```diff
@@ -136,14 +178,36 @@ impl Iterator for TokenIter<'a> {
     }
 }
 
-fn get_real_ident_class(text: &str, edition: Edition) -> Class {
-    match text {
+/// Classifies into identifier class; returns `None` if this is a non-keyword identifier.
+fn get_real_ident_class(text: &str, edition: Edition, allow_path_keywords: bool) -> Option<Class> {
+    let ignore: &[&str] =
+        if allow_path_keywords { &["self", "Self", "super", "crate"] } else { &["self", "Self"] };
+    if ignore.iter().any(|k| *k == text) {
+        return None;
+    }
+    Some(match text {
         "ref" | "mut" => Class::RefKeyWord,
-        "self" | "Self" => Class::Self_,
         "false" | "true" => Class::Bool,
         _ if Symbol::intern(text).is_reserved(|| edition) => Class::KeyWord,
-        _ => Class::Ident,
-    }
+        _ => return None,
+    })
+}
+
+/// Before explaining what this function does, some global explanations on rust's `Span`:
+///
+/// Each source code file is stored in the source map in the compiler and has a
+/// `lo` and a `hi` (lowest and highest bytes in this source map which can be seen as one huge
+/// string to simplify things). So in this case, this represents the starting byte of the current
+/// file. It'll be used later on to retrieve the "definition span" from the
+/// `span_correspondance_map` (which is inside `context`).
+///
+/// This when we transform the "span" we have from reading the input into a "span" which can be
+/// used as index to the `span_correspondance_map` to get the definition of this item.
+///
+/// So in here, `file_span_lo` is representing the "lo" byte in the global source map, and to make
+/// our "span" works in there, we simply add `file_span_lo` to our values.
+fn local_span_to_global_span(file_span_lo: u32, start: u32, end: u32) -> (u32, u32) {
+    (start + file_span_lo, end + file_span_lo)
 }
```

Comment:
This function makes me very nervous, there's nothing preventing you from calling it twice or forgetting to call it. Could we use proper `Span`s instead?

Reply:
I'd rather not. The operations they perform are much more "expensive". However I can suggest the following:

The `span_map` will use a `LightSpan` type (which will be `lo` and `hi` and will implement a `new` method which will take the `file_span_lo` so it's not possible to forget about it.